filebeat/input/filestream/internal/input-logfile/metrics.go

```diff
@@ -54,17 +55,29 @@ func (m *Metrics) Close() {
 	m.unregister()
 }
 
-func NewMetrics(id string) *Metrics {
+func NewMetrics(ctx v2.Context, id string) *Metrics {
 	// The log input creates the `filebeat.harvester` registry as a package
 	// variable, so it should always exist before this function runs.
-	// However at least on testing scenarios this does not hold true, so
+	// However, at least on testing scenarios this does not hold true, so
 	// if needed, we create the registry ourselves.
 	harvesterMetrics := monitoring.Default.GetRegistry("filebeat.harvester")
 	if harvesterMetrics == nil {
 		harvesterMetrics = monitoring.Default.NewRegistry("filebeat.harvester")
 	}
 
-	reg, unreg := inputmon.NewInputRegistry("filestream", id, nil)
+	var reg *monitoring.Registry
+	var unreg = func() {}
+	if id != "" {
+		reg = ctx.EnhanceMetricRegistry(id, "filestream")
+		err := inputmon.RegisterMetrics(id, reg)
+		if err != nil {
+			ctx.Logger.Warn("failed to register filestream metrics: ", err)
+		}
+		unreg = func() { inputmon.UnregisterMetrics(id) }
```

Comment:
Unregistering is handled by the pipeline now, right? (In `compat.go` below.) So we should be able to remove `unreg` entirely.

Reply:
+1

Reply:
I don't think these things should be separate... if they have to be handled via different helpers due to package constraints, I still don't think we should expose that separation to the consumers of the v2 input API, if we can tell when the input is shutting down that should be enough and we can handle whatever cancellation is needed at that point instead of making every input responsible for extra shutdown boilerplate.